### ISSUE TITLE
prepare_file for configs in tests doesnt work

### ISSUE THREAD

**Issue by AbdealiJK** [coala-analyzer/coala-bears#110]
The issue here is that the `with` statement is wrapping the creation of the class. Not the running of the class.

So, when we do 
```
with prepare_file(config_file,
                  filename=None,
                  force_linebreaks=True,
                  create_tempfile=True) as (conf_lines, conf_file):
    JSHintBearConfigFileTest = verify_local_bear(
        JSHintBear,
        valid_files=(),
        invalid_files=(test_file1, test_file2),
        settings={"jshint_config": conf_file})
```
The file in `config_file` exists only when pytest is collecting for the unittests to run. NOT when it is actually running the unittest.

Not sure how we fix this. Possible solution in using `pytest`'s fixtures.
The other (not so nice) solution is again use `test_files` and create config files manually.

**Comment by tushar-rishav** [comment]
@AbdealiJK Could you elaborate "collecting for the unittests to run...actually running the unittest" ? Are you implying that py.test performs two step process. First collecting all the tests files and then actually run the tests?

**Comment by AbdealiJK** [comment]
@tushar-rishav yes, it is a two step process

**Comment by tushar-rishav** [comment]
@AbdealiJK Cool. Could you elaborate why py.test is ignoring the config_file while running the test? Umm is py.test sort of caching the test files in first step and then using those in second step? I guess not, because that would not be the cause for missing data from config_file. Some more thoughts please?

**Comment by AbdealiJK** [comment]
So, in the first step it just goes through all files and checks which classes are children of `unittest.TestCase`. Once it collects all these classes it uses them one by one and runs them. It must be creating objects out of the classes i guess.
The thing is that when it is collecting the classes, the `with` statement where the `config_file` is created has already completed executing. So, the temp config file which was created has already been deleted when the object is being run - As the control went out of the `with` block.

That's why the config file doesnt exist when the object is being run later.

**Comment by tushar-rishav** [comment]
@AbdealiJK Ah If py.test stores the instances then how about if we bind the conf_file to the class itself and pass that class variable as an argument to settings. Something like
```python
TempClass = object()
with prepare_file(config_file,
                  filename=None,
                  force_linebreaks=True,
                  create_tempfile=True) as (conf_lines, conf_file):
    TempClass = verify_local_bear(
        JSHintBear,
        valid_files=(test_file1, test_file2),
        invalid_files=(),
        settings={"jshint_config": conf_file})
    TempClass.conf_file = conf_file
    JSHintBearConfigFileTest = verify_local_bear(
        JSHintBear,
        valid_files=(test_file1, test_file2),
        invalid_files=(),
        settings={"jshint_config": TempClass.conf_file})

```
Might work? Umm the bad thing is we are calling `verify_local_bear` twice.

**Comment by AbdealiJK** [comment]
I do not see how that will help ... `conf_file` is just a string ... the file would still get deleted

**Comment by tushar-rishav** [comment]
@AbdealiJK Yes it's just the name of the file. So will the files be deleted even if it's reference count is non zero? I guess when first time `with`block is executed, we get a temp file path and lines. Now once the block is executed the temp file would be deleted because it's not being referenced anywhere. My doubt is will the files get deleted even if we create a new reference to it? I think unless ref count is zero  garbage collector won't touch any object on it's own. Umm right?

**Comment by AbdealiJK** [comment]
No, that's not how `prepare_file()` works. We wrote that btw (You can check out ContextManagers)
We delete the file in the `finally` - i.e. when the control gets out of the `with` block.

Also, as I mentioned before, the `conf_file` variable is just a string. There are no reference counters for strings in python. Strings are immutable types in python.

**Comment by Makman2** [comment]
ha I was right! Occurred in some PR too, so can confirm that :+1:

**Comment by Makman2** [comment]
So the only solution is to move the functionality into `verify_local_bear` that takes care of that inside `setUp()` and `tearDown()`.
